incubator/virtualcluster/pkg/syncer/resources/pod/dws.go

```diff
@@ -138,78 +138,66 @@ func (c *controller) reconcilePodCreate(cluster, namespace, name string, vPod *v
-	if pPod.Spec.ServiceAccountName != "" {
-		saName = pPod.Spec.ServiceAccountName
-	}
-
-	// find service account token secret and replace the one set by tenant kcm.
-	secretList, err := c.client.Secrets(targetNamespace).List(metav1.ListOptions{
-		LabelSelector: labels.Set(map[string]string{
-			constants.LabelServiceAccountName: saName,
-		}).String(),
-	})
-	if err != nil && !errors.IsNotFound(err) {
-		return err
-	}
-
-	if secretList == nil || len(secretList.Items) == 0 {
-		return fmt.Errorf("service account token secret for pod is not ready")
-	}
-	pSecret := &secretList.Items[0]
-
-	tenantClient, err := c.multiClusterPodController.GetClusterClient(cluster)
+	pSecret, err := c.getPodServiceAccountSecret(pPod)
 	if err != nil {
-		return err
-	}
-	vSecret, err := utils.GetSecret(tenantClient.CoreV1(), namespace, saName)
-	if err != nil {
-		return fmt.Errorf("failed to get secret: %v", err)
+		return fmt.Errorf("failed to get service account secret from cluster %s cache: %v", cluster, err)
 	}
 
 	services, err := c.getPodRelatedServices(cluster, pPod)
 	if err != nil {
 		return fmt.Errorf("failed to list services from cluster %s cache: %v", cluster, err)
 	}
 
-	if len(services) == 0 {
-		return fmt.Errorf("service is not ready")
-	}
-
-	nameServer, err := c.getClusterNameServer(c.client, cluster)
+	nameServer, err := c.getClusterNameServer(cluster)
 	if err != nil {
 		return fmt.Errorf("failed to find nameserver: %v", err)
 	}
 
 	var ms = []conversion.PodMutator{
-		conversion.PodMutateDefault(vPod, vSecret, pSecret, services, nameServer),
+		conversion.PodMutateDefault(vPod, pSecret, services, nameServer),
 		conversion.PodMutateAutoMountServiceAccountToken(c.config.DisableServiceAccountToken),
-		conversion.PodAddExtensionMeta(vPod),
+		// TODO: make extension configurable
+		//conversion.PodAddExtensionMeta(vPod),
 	}
 
 	err = conversion.VC(c.multiClusterPodController, cluster).Pod(pPod).Mutate(ms...)
 	if err != nil {
 		return fmt.Errorf("failed to mutate pod: %v", err)
 	}
-
 	_, err = c.client.Pods(targetNamespace).Create(pPod)
 	if errors.IsAlreadyExists(err) {
 		klog.Infof("pod %s/%s of cluster %s already exist in super master", namespace, name, cluster)
 		return nil
 	}
+
 	return err
 }
 
-func (c *controller) getClusterNameServer(client v1core.ServicesGetter, cluster string) (string, error) {
-	svc, err := client.Services(conversion.ToSuperMasterNamespace(cluster, constants.TenantDNSServerNS)).Get(constants.TenantDNSServerServiceName, metav1.GetOptions{})
+func (c *controller) getPodServiceAccountSecret(pPod *v1.Pod) (*v1.Secret, error) {
+	saName := "default"
+	if pPod.Spec.ServiceAccountName != "" {
+		saName = pPod.Spec.ServiceAccountName
+	}
+	// find service account token secret and replace the one set by tenant kcm.
+	req, _ := labels.NewRequirement(constants.LabelServiceAccountName, selection.In, []string{saName})
+	secretList, err := c.secretLister.Secrets(pPod.Namespace).List(labels.NewSelector().Add(*req))
```

Comment:
maybe you want `labels.SelectorFromSet(map[string]string{constants.LabelServiceAccountName: saName})`